app/migrations.go

```diff
@@ -156,3 +157,78 @@ func (a *App) DoEmojisPermissionsMigration() {
 		mlog.Critical(fmt.Sprint(result.Err))
 	}
 }
+
+func (a *App) DoGuestRolesCreationMigration() {
+	// If the migration is already marked as completed, don't do it again.
+	if result := <-a.Srv.Store.System().GetByName(GUEST_ROLES_CREATION_MIGRATION_KEY); result.Err == nil {
+		return
+	}
+
+	roles := model.MakeDefaultRoles()
+
+	allSucceeded := true
+	if result := <-a.Srv.Store.Role().GetByName(model.CHANNEL_GUEST_ROLE_ID); result.Err != nil {
+		if result := <-a.Srv.Store.Role().Save(roles[model.CHANNEL_GUEST_ROLE_ID]); result.Err != nil {
+			mlog.Critical("Failed to create new guest role to database.")
+			mlog.Critical(fmt.Sprint(result.Err))
+			allSucceeded = false
+		}
+	}
+	if result := <-a.Srv.Store.Role().GetByName(model.TEAM_GUEST_ROLE_ID); result.Err != nil {
+		if result := <-a.Srv.Store.Role().Save(roles[model.TEAM_GUEST_ROLE_ID]); result.Err != nil {
+			mlog.Critical("Failed to create new guest role to database.")
+			mlog.Critical(fmt.Sprint(result.Err))
+			allSucceeded = false
+		}
+	}
+	if result := <-a.Srv.Store.Role().GetByName(model.SYSTEM_GUEST_ROLE_ID); result.Err != nil {
+		if result := <-a.Srv.Store.Role().Save(roles[model.SYSTEM_GUEST_ROLE_ID]); result.Err != nil {
+			mlog.Critical("Failed to create new guest role to database.")
+			mlog.Critical(fmt.Sprint(result.Err))
+			allSucceeded = false
+		}
+	}
+
+	resultSchemes := <-a.Srv.Store.Scheme().GetAllPage("", 0, 1000000)
+	if resultSchemes.Err != nil {
+		mlog.Critical("Failed to get all schemes.")
+		mlog.Critical(fmt.Sprint(resultSchemes.Err))
+		allSucceeded = false
+	}
+	schemes := resultSchemes.Data.([]*model.Scheme)
+	for _, scheme := range schemes {
+		if scheme.DefaultTeamGuestRole == "" || scheme.DefaultChannelGuestRole == "" {
+			scheme.DefaultTeamGuestRole = model.TEAM_GUEST_ROLE_ID
```

Comment:
I don't think this is safe. You need to create roles that are copies of the default guest roles, one per scheme, and assign them to the individual schemes. Otherwise you could break things if you edit guest permissions for one scheme and it will affect the global default scheme and other schemes.

The new roles need to be created just like how they are created in the `CreateScheme` store function.

Reply:
Done! :)